### PR TITLE
Adding github action workflows for kind

### PR DESCRIPTION
**Pull request by agiwalpooja20**
Creating a workflow to deploy

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/1625)

### CODE REVIEW

File: .github/workflows/gke-workflow.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        with:
+          go-version: 1.12
+
+      - name: Install correct helm version
+        run: |
+          helm version
+          curl -LO https://git.io/get_helm.sh
+          chmod 700 get_helm.sh
+          ./get_helm.sh -v v2.16.6
+          kubectl create serviceaccount --namespace kube-system tiller
```
Comment (vishal-biyani) [inline .github/workflows/gke-workflow.yaml:58]: Tiller SA will go away for Helm3

File: .github/workflows/kind-workflow.yaml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    branches:
+      - fission-github-actions
+    types: [labeled]
+jobs:
+  fission-on-kind:
+    runs-on: ubuntu-latest
+    if: github.event.label.name == 'kind'
+    steps:
+      - name: Free up some space
+        run: |
```
Comment (vishal-biyani) [inline .github/workflows/kind-workflow.yaml:13]: Just trying to understand why these two commands are needed?

File: .github/workflows/kind-workflow.yaml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+      - name: Checkout fisson from master
+        uses: actions/checkout@v2
+        with:
+          ref: fission-github-actions
+          
+      - name: Install kind cluster with registry
+        run: |
+          cd /home/runner/work/fission/fission 
+          ./kind-with-registry.sh
```
Comment (vishal-biyani) [inline .github/workflows/kind-workflow.yaml:25]: Can we please move this script inside test folder?

File: .github/workflows/kind-workflow.yaml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        uses: actions/setup-go@v1
+        with:
+          go-version: 1.12
+
+      - name: Install correct helm version
+        run: |
+          helm version
+          curl -LO https://git.io/get_helm.sh
+          chmod 700 get_helm.sh
+          ./get_helm.sh -v v2.16.6
```
Comment (vishal-biyani) [inline .github/workflows/kind-workflow.yaml:56]: Let's use Helm3 instead of Helm2

File: .github/workflows/kind-workflow.yaml [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+          kubectl create serviceaccount --namespace kube-system tiller
```
Comment (vishal-biyani) [inline .github/workflows/kind-workflow.yaml:57]: Once Helm3 is used, these 3 statements will not be needed

File: environments/python/Dockerfile-2.7 [hunk 6]
```diff
@@ -2,7 +2,7 @@ FROM alpine:3.5
 
 WORKDIR /app
 
-RUN apk add --no-cache python python-dev build-base py-pip libev-dev libffi-dev && \
+RUN apk add --no-cache python python-dev build-base py-pip libev-dev libffi-dev  gcc bash py-pip && \
```
Comment (vishal-biyani) [inline environments/python/Dockerfile-2.7:5]: Sorry, why this change is required? Ideally, the environment image should not have to change at all

File: environments/python/Dockerfile [hunk 7]
```diff
@@ -2,7 +2,9 @@ FROM alpine
 
 WORKDIR /app
 
-RUN apk add --no-cache python3 python3-dev build-base libev-dev libffi-dev && \
+RUN apk add --update gcc && \
```
Comment (vishal-biyani) [inline environments/python/Dockerfile:5]: Sorry, why this change is required? Ideally, the environment image should not have to change at all

File: environments/python/builder/Dockerfile [hunk 8]
```diff
@@ -2,7 +2,7 @@ ARG BUILDER_IMAGE=fission/builder:latest
 FROM ${BUILDER_IMAGE}
 
 RUN apk update
-RUN apk add --no-cache python3 python3-dev build-base
+RUN apk add --no-cache python3 python3-dev gcc build-base bash py-pip
```
Comment (vishal-biyani) [inline environments/python/builder/Dockerfile:5]: Sorry, why this change is required? Ideally, the environment image should not have to change at all

File: kind-with-registry.sh [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+fi
+
+reg_host="${reg_name}"
+if [ "${kind_network}" = "bridge" ]; then
+    reg_host="$(docker inspect -f '{{.NetworkSettings.IPAddress}}' "${reg_name}")"
+fi
+echo "Registry Host: ${reg_host}"
+
+# create a cluster with the local registry enabled in containerd
+cat <<EOF | kind create cluster --name "${KIND_CLUSTER_NAME}" --config=-
```
Comment (vishal-biyani) [inline kind-with-registry.sh:50]: It will be good to have a way to configure based on a flag - a single node or multi node cluster

File: kind-with-registry.sh-BAK [hunk 10]
```diff
@@ -0,0 +1,40 @@
+#!/bin/sh
```
Comment (vishal-biyani) [inline kind-with-registry.sh-BAK:1]: I am assuming this BAK file will be removed

File: skaffold.yaml [hunk 11]
```diff
@@ -3,7 +3,7 @@
 # build and deploy Fission to Kubernetes cluster.
 ############## Usage ##############
 # Skaffold CLI should be installed on your machine. 
-# Replace <DOCKERHUB_REPO> with your registry repo/org name
+# Replace fission with your registry repo/org name
```
Comment (vishal-biyani) [inline skaffold.yaml:6]: This change is not needed

File: skaffold.yaml [hunk 12]
```diff
@@ -12,6 +12,12 @@
 apiVersion: skaffold/v2beta1
 kind: Config
 build:
+  local:
+    useDockerCLI: false
+    useBuildkit: false
+    push: true
+  insecureRegistries:
```
Comment (vishal-biyani) [inline skaffold.yaml:19]: I think we should have 1/2 different profiles in Skaffold. Because as a developer I may not want to have a registry locally. In case of GKE workflow - this not work at all. And in case of Kind this is needed.

File: test/kind_build.sh [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+dump_system_info
+
+setupCIBuildEnv
+
+build_and_push_go_mod_cache_image $REPO/go-mod-image-cache
+build_and_push_fission_bundle $REPO/$IMAGE:$TAG $REPO/go-mod-image-cache
+build_and_push_pre_upgrade_check_image $PRE_UPGRADE_CHECK_IMAGE:$TAG $REPO/go-mod-image-cache
+build_and_push_fetcher $FETCHER_IMAGE:$TAG $REPO/go-mod-image-cache
+build_and_push_builder $BUILDER_IMAGE:$TAG $REPO/go-mod-image-cache
```
Comment (vishal-biyani) [inline test/kind_build.sh:28]: Since we are now using Skaffold - bundle, fetcher and pre-upgrade still needs to be built?

File: test/kind_test.sh [hunk 14]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+set -euo pipefail
+
+# Unbound variables cause failure, so this readable if block instead of Parameter Expansion
+if [[ ${TRAVIS_EVENT_TYPE+NOVALUE} == "cronNOVALUE" ]]
```
Comment (vishal-biyani) [inline test/kind_test.sh:6]: I think the Travis variable is of no use anymore?

File: test/kind_test_utils.sh [hunk 15]
```diff
@@ -0,0 +1,645 @@
...
+# Test runner. Shell scripts that build fission CLI and server, push a
+# docker image to GCR, deploy it on a cluster, and run tests against
+# that deployment.
+#
+
+set -euo pipefail
+
+ROOT=`realpath $(dirname $0)/..`
+
+travis_fold_start() {
```
Comment (vishal-biyani) [inline test/kind_test_utils.sh:13]: Please remove all `Travis` references

File: test/kind_test_utils.sh [hunk 16]
```diff
@@ -0,0 +1,645 @@
...
+}
+
+build_and_push_go_mod_cache_image() {
+    image_tag=$1
+    travis_fold_start go_mod_cache_image $image_tag
+
+#    gcloud_login
+
+    if ! gcloud docker -- pull $image_tag >/dev/null 2>&1 ; then
+      docker build -q -t $image_tag -f $ROOT/cmd/fission-bundle/Dockerfile.fission-bundle --target godep --build-arg GITCOMMIT=$(getGitCommit) --build-arg BUILDDATE=$(getDate) --build-arg BUILDVERSION=$(getVersion) .
```
Comment (vishal-biyani) [inline test/kind_test_utils.sh:93]: Is this still required with Skaffold in place?

File: test/kind_test_utils.sh [hunk 17]
```diff
@@ -0,0 +1,645 @@
...
+    else
+      docker build -q -t $image_tag -f $ROOT/cmd/fission-bundle/Dockerfile.fission-bundle --cache-from ${image_tag} --target godep --build-arg GITCOMMIT=$(getGitCommit) --build-arg BUILDDATE=$(getDate) --build-arg BUILDVERSION=$(getVersion) .
+    fi
+
+#    gcloud docker -- push $image_tag &
+    docker push $image_tag &
+    travis_fold_end go_mod_cache_image
+}
+
+build_and_push_pre_upgrade_check_image() {
```
Comment (vishal-biyani) [inline test/kind_test_utils.sh:103]: Duplicate of Skaffold work?

File: test/kind_test_utils.sh [hunk 18]
```diff
@@ -0,0 +1,645 @@
...
+    helm dependency update $ROOT/charts/fission-all
+
+    echo "Installing fission"
+    helm install		\
+	 --wait			\
+	 --timeout 540	        \
+	 --name $id		\
+	 --set $helmVars	\
+	 --namespace $ns        \
+	 $ROOT/charts/fission-all
```
Comment (vishal-biyani) [inline test/kind_test_utils.sh:278]: Duplicate of Skaffold work?

File: test/kind_test_utils.sh [hunk 19]
```diff
@@ -0,0 +1,645 @@
...
+    echo "--- end kubectl events $ns ---"
+}
+export -f dump_kubernetes_events
+
+dump_tiller_logs() {
+    echo "--- tiller logs ---"
+    tiller_pod=`kubectl get pods -n kube-system | grep tiller| tr -s " "| cut -d" " -f1`
+    kubectl logs $tiller_pod --since=30m -n kube-system
+    echo "--- end tiller logs ---"
+}
```
Comment (vishal-biyani) [inline test/kind_test_utils.sh:303]: Remove with Helm3

File: test/kind_test_utils.sh (hunk 16 above)
Comment (agiwalpooja20) [reply test/kind_test_utils.sh:93]: @vishal-biyani Changed this because it was giving error for pip3 command not found
```/bin/sh: pip3: not found
The command '/bin/sh -c apk add --no-cache python3 python3-dev build-base libev-dev libffi-dev &&     pip3 install --upgrade pip &&     rm -r /root/.cache' returned a non-zero code: 127
##[error]Process completed with exit code 127.```

File: environments/python/Dockerfile-2.7 (hunk 6 above)
Comment (agiwalpooja20) [reply environments/python/Dockerfile-2.7:5]: @vishal-biyani Changed this because it was giving error for pip3 command not found
```/bin/sh: pip3: not found
The command '/bin/sh -c apk add --no-cache python3 python3-dev build-base libev-dev libffi-dev &&     pip3 install --upgrade pip &&     rm -r /root/.cache' returned a non-zero code: 127
##[error]Process completed with exit code 127.```

Comment (vishal-biyani) [reply environments/python/Dockerfile-2.7:5]: @agiwalpooja20 So we are facing same issue in all our builds right now - but this should be fixed as a separate PR anyway. Plus I am not able to reproduce locally when I build the same image - so there is something else going on there.

File: .github/workflows/kind-workflow.yaml (hunk 2 above)
Comment (abhisheks-infracloud) [reply .github/workflows/kind-workflow.yaml:13]: This was added because sometime GitHub runner's storage gets full and it gives `no space left on device error`. GitHub runner are configured with limited resources as mentioned in the official document. 
https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources

Issue raised on GitHub:
https://github.com/actions/virtual-environments/issues/709

**Comment by vishal-biyani** [comment]
Closing this as the Kind PR is merged now #1876
